test/unit_test/pmf_to_pdf_test.cc

```diff
@@ -0,0 +1,130 @@
+#define BOOST_TEST_DYN_LINK
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+#include "learner.h"
+#include "pmf_to_pdf.h"
+#include "parse_args.h"  // setup_base()
+#include "action_score.h"
+
+using namespace VW::LEARNER;
+using namespace std;
+
+
+namespace VW { namespace pmf_to_pdf {
+
+  void learn(VW::pmf_to_pdf::reduction& data, single_learner& base, example& ec);
+  void predict(VW::pmf_to_pdf::reduction& data, single_learner& base, example& ec);
+
+  struct reduction_test_harness
+  {
+    reduction_test_harness() : _curr_idx(0) {}
+
+    void set_predict_response(const vector<pair<uint32_t, float>>& predictions) { _predictions = predictions; }
+
+    void test_predict(single_learner& base, example& ec)
+    {
+      ec.pred.a_s.clear();
+      for (uint32_t i = 0; i < _predictions.size(); i++)
+      {
+        ec.pred.a_s.push_back(ACTION_SCORE::action_score{_predictions[i].first, _predictions[i].second});
+      }
+
+      cout << "\nec.pred.a_s (PMF): " << endl;
+      for (uint32_t i = 0; i < _predictions.size(); i++)
+      {
+        cout << "(" << ec.pred.a_s[i].action << " : " << ec.pred.a_s[i].score << "), " << endl;
+      }
+    }
+
+    void test_learn(single_learner& base, example& ec)
+    {
+      cout << "ec.l.cb.costs after:" << endl;
+      for (uint32_t i = 0; i < ec.l.cb.costs.size(); i++)
+      {
+        cout << "(" << ec.l.cb.costs[i].action << " , " << ec.l.cb.costs[i].cost << " , " << ec.l.cb.costs[i].probability
+          << " , " << ec.l.cb.costs[i].partial_prediction << "), " << endl;
+      }
+    }
+
+    static void predict(reduction_test_harness& test_reduction, single_learner& base, example& ec)
+    {
+      test_reduction.test_predict(base, ec);
+    }
+
+    static void learn(reduction_test_harness& test_reduction, single_learner& base, example& ec)
+    {
+      test_reduction.test_learn(base, ec);
+    };
+
+  private:
+    vector<pair<uint32_t, float>> _predictions;
+    int _curr_idx;
+  };
+
+  using test_learner_t = learner<reduction_test_harness, example>;
+  using predictions_t = vector<pair<uint32_t, float>>;
+
+  test_learner_t* get_test_harness_reduction(const predictions_t& base_reduction_predictions);
+
+}} //namespace
+
+BOOST_AUTO_TEST_CASE(pmf_to_pdf_basic)
+{
```

Comment:
Should this test be `asserting` various states instead of printing to `cout`?

Reply:
Yes. Fixed.